index.js

```diff
@@ -104,51 +77,279 @@ module.exports = {
+      responseHandler();
+    } else if (jso.error !== undefined) {
+      errorHandler();
+    } else {
+      this.configuration.errorHandler(new ErrorWithData("Received an invalid JSON-RPC message.", jso));
+    }
   },
 
-  disableHostedNodeFallback: function () {
-    this.useHostedNodeFallback = false;
-    this.nodes.hosted = [];
-    if (this.wsUrl === HOSTED_WEBSOCKET) {
-      this.wsUrl = null;
+  /**
+   * Used internally.  Setup the initial subscription for new blocks.  Called on first connect, then never again.
+   * 
+   * @param {function():void} callback - called when the subscription is done being setup
+   */
+  setupBlockSubscription: function (callback) {
+    var handleSubscribeToNewHeadsResult = function(resultOrError) {
+      if (resultOrError instanceof Error || resultOrError.error) {
+        this.internalState.newBlockIntervalTimeoutId = setInterval(this.getBlockByNumber.bind(this, "latest", false, this.onNewBlock.bind(this, true)), this.BLOCK_POLL_INTERVAL);
+      } else {
+        this.internalState.subscriptions[resultOrError] = this.onNewBlock.bind(this, true);
+      }
+      callback();
+      callback = function () { };
+    };
+
+    var subscribeToNewBlocks = function(sync) {
+      var maybeCallback = (sync) ? undefined : handleSubscribeToNewHeadsResult.bind(this);
+      var result = this.subscribeNewHeads(maybeCallback);
+      if (sync) handleSubscribeToNewHeadsResult(result);
+    }.bind(this);
+
+    this.internalState.transporter.addReconnectListener(subscribeToNewBlocks.bind(this, false));
+    clearInterval(this.internalState.newBlockIntervalTimeoutId);
+    subscribeToNewBlocks(!callback);
+  },
+
+  onNewBlock: function (isLatest, block) {
+    if (typeof isLatest !== "boolean") throw new Error("isLatest must be a boolean");
+    if (typeof block !== "object") throw new Error("block must be an object");
+
+    // adjust the seen block numbers
+    var newBlockNumber = parseInt(block.number, 16);
+    if (this.internalState.lowestSeenSequentialBlock === null)
+      this.internalState.lowestSeenSequentialBlock = newBlockNumber;
+    if (this.internalState.lastSeenLatestBlockNumber === null)
+      this.internalState.lastSeenLatestBlockNumber = newBlockNumber;
+    if (isLatest) {
+      this.internalState.lastSeenLatestBlockNumber = newBlockNumber;
+      this.block = block;
+    }
+    if (this.internalState.lowestSeenSequentialBlock + 1 === newBlockNumber)
+      this.internalState.lowestSeenSequentialBlock = newBlockNumber;
+
+    // FIXME: ethrpc should really store the original block and add getters for making it easier to interact with
+    block.number = newBlockNumber;
+
+    // fetch next missing block if necessary (this will result in recursively catching up)
+    if (this.internalState.lowestSeenSequentialBlock != this.internalState.lastSeenLatestBlockNumber)
+      this.getBlockByNumber(this.internalState.lowestSeenSequentialBlock + 1, false, this.onNewBlock.bind(this, false));
```

Comment:
> However, if you think we'll be taking a different route for missed blocks (or perhaps we just don't care about missed blocks) then I suspect a lot of this code could be deleted and replaced with something much simpler like renaming onNewBlock to onNewLatestBlock and not calling it for anything other than latest blocks.

Let's do this for now.  If we end up wanting to use this for backfills, this code is now saved via git so we can copy-and-paste it easily enough if needed.